# **1. Install Dependencies**

In [48]:
import numpy as np
import pandas as pd

# for visualization
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

# for ignore the Warnings
import warnings
warnings.filterwarnings("ignore")

# **2. Import the Datasets**

In [49]:
df = pd.read_csv('Employee_attrition_dataset.csv')
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   
1                        3    Male          61               2         2   
2                        4    Male          92               2         1   
3                        4  Female          56               3         1   
4                        1    Male          40               3         1   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  \
0        Sales Executive                4        Single           5993   
1     Research Scientist                2       Married           5130   
2  Laboratory Technician                3        Single           2090   
3     Research Scientist                3       Married           2909   
4  Laboratory Technician                2       Married           3468   

   MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
0        19479                   8      Y      Yes                 11   
1        24907                   1      Y       No                 23   
2         2396                   6      Y      Yes                 15   
3        23159                   1      Y      Yes                 11   
4        16632                   9      Y       No                 12   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
0                  3                         1             80   
1                  4                         4             80   
2                  3                         2             80   
3                  3                         3             80   
4                  3                         4             80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

# **3. Undestanding The Data**

In [50]:
# showing 10 random samples
df.sample(10)

Age Attrition     BusinessTravel  DailyRate              Department  \
214   30       Yes      Travel_Rarely       1005  Research & Development   
323   28       Yes      Travel_Rarely       1157  Research & Development   
923   34        No  Travel_Frequently        648         Human Resources   
45    41       Yes      Travel_Rarely       1360  Research & Development   
816   39        No         Non-Travel        439  Research & Development   
400   39        No  Travel_Frequently       1218  Research & Development   
637   32        No         Non-Travel        300  Research & Development   
766   50        No      Travel_Rarely       1464  Research & Development   
752   36       Yes      Travel_Rarely        885  Research & Development   
47    37        No      Travel_Rarely        408  Research & Development   

     DistanceFromHome  Education    EducationField  EmployeeCount  \
214                 3          3  Technical Degree              1   
323                 2          4           Medical              1   
923                11          3     Life Sciences              1   
45                 12          3  Technical Degree              1   
816                 9          3     Life Sciences              1   
400                 1          1     Life Sciences              1   
637                 1          3     Life Sciences              1   
766                 2          4           Medical              1   
752                16          4     Life Sciences              1   
47                 19          2     Life Sciences              1   

     EmployeeNumber  EnvironmentSatisfaction  Gender  HourlyRate  \
214             297                        4  Female          88   
323             440                        1    Male          84   
923            1289                        3    Male          56   
45               58                        2  Female          49   
816            1132                        3    Male          70   
400             531                        2    Male          52   
637             882                        4    Male          61   
766            1061                        2    Male          62   
752            1042                        3  Female          43   
47               61                        2    Male          73   

     JobInvolvement  JobLevel                JobRole  JobSatisfaction  \
214               3         1     Research Scientist                1   
323               1         1     Research Scientist                4   
923               2         2        Human Resources                2   
45                3         5      Research Director                3   
816               3         2  Laboratory Technician                2   
400               3         5                Manager                3   
637               3         1  Laboratory Technician                4   
766               3         5      Research Director                3   
752               4         1  Laboratory Technician                1   
47                3         1     Research Scientist                2   

    MaritalStatus  MonthlyIncome  MonthlyRate  NumCompaniesWorked Over18  \
214        Single           2657         8556                   5      Y   
323       Married           3464        24737                   5      Y   
923       Married           4490        21833                   4      Y   
45        Married          19545        16280                   1      Y   
816        Single           6782         8770                   9      Y   
400      Divorced          19197         8213                   1      Y   
637      Divorced           2314         9148                   0      Y   
766       Married          19237        12853                   2      Y   
752        Single           2743         8269                   1      Y   
47        Married           3022        10227                   4      Y   

    OverTime  Perc

* Target column is Attrition in this dataset

In [51]:
# shape of the data
df.shape

(1470, 35)

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [53]:
# columns present in the data
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

In [54]:
df.dtypes

Age                          int64
Attrition                   object
BusinessTravel              object
DailyRate                    int64
Department                  object
DistanceFromHome             int64
Education                    int64
EducationField              object
EmployeeCount                int64
EmployeeNumber               int64
EnvironmentSatisfaction      int64
Gender                      object
HourlyRate                   int64
JobInvolvement               int64
JobLevel                     int64
JobRole                     object
JobSatisfaction              int64
MaritalStatus               object
MonthlyIncome                int64
MonthlyRate                  int64
NumCompaniesWorked           int64
Over18                      object
OverTime                    object
PercentSalaryHike            int64
PerformanceRating            int64
RelationshipSatisfaction     int64
StandardHours                int64
StockOptionLevel             int64
TotalWorkingYears   

In [55]:
# numerical data presnt in the dataset
num_data = df.select_dtypes(include=["int64","float64"])
print("numerical columns: ",num_data.columns)

numerical columns:  Index(['Age', 'DailyRate', 'DistanceFromHome', 'Education', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'MonthlyIncome',
       'MonthlyRate', 'NumCompaniesWorked', 'PercentSalaryHike',
       'PerformanceRating', 'RelationshipSatisfaction', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole',
       'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')


In [56]:
# categorical data present in the dataset
cat_data = df.select_dtypes(include=['object'])
print("categorical columns: ",cat_data.columns)

categorical columns:  Index(['Attrition', 'BusinessTravel', 'Department', 'EducationField', 'Gender',
       'JobRole', 'MaritalStatus', 'Over18', 'OverTime'],
      dtype='object')


In [57]:
for i in cat_data:
    print(i,df[i].unique())
    print(df[i].value_counts())
    print("***************************")

Attrition ['Yes' 'No']
No     1233
Yes     237
Name: Attrition, dtype: int64
***************************
BusinessTravel ['Travel_Rarely' 'Travel_Frequently' 'Non-Travel']
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: BusinessTravel, dtype: int64
***************************
Department ['Sales' 'Research & Development' 'Human Resources']
Research & Development    961
Sales                     446
Human Resources            63
Name: Department, dtype: int64
***************************
EducationField ['Life Sciences' 'Other' 'Medical' 'Marketing' 'Technical Degree'
 'Human Resources']
Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: EducationField, dtype: int64
***************************
Gender ['Female' 'Male']
Male      882
Female    588
Name: Gender, dtype: int64
***************************
JobRole ['Sales Executive' 'Research Scientist' 'Laboratory

In [58]:
df.describe().T

count          mean          std     min      25%  \
Age                       1470.0     36.923810     9.135373    18.0    30.00   
DailyRate                 1470.0    802.485714   403.509100   102.0   465.00   
DistanceFromHome          1470.0      9.192517     8.106864     1.0     2.00   
Education                 1470.0      2.912925     1.024165     1.0     2.00   
EmployeeCount             1470.0      1.000000     0.000000     1.0     1.00   
EmployeeNumber            1470.0   1024.865306   602.024335     1.0   491.25   
EnvironmentSatisfaction   1470.0      2.721769     1.093082     1.0     2.00   
HourlyRate                1470.0     65.891156    20.329428    30.0    48.00   
JobInvolvement            1470.0      2.729932     0.711561     1.0     2.00   
JobLevel                  1470.0      2.063946     1.106940     1.0     1.00   
JobSatisfaction           1470.0      2.728571     1.102846     1.0     2.00   
MonthlyIncome             1470.0   6502.931293  4707.956783  1009.0  2911.00   
MonthlyRate               1470.0  14313.103401  7117.786044  2094.0  8047.00   
NumCompaniesWorked        1470.0      2.693197     2.498009     0.0     1.00   
PercentSalaryHike         1470.0     15.209524     3.659938    11.0    12.00   
PerformanceRating         1470.0      3.153741     0.360824     3.0     3.00   
RelationshipSatisfaction  1470.0      2.712245     1.081209     1.0     2.00   
StandardHours             1470.0     80.000000     0.000000    80.0    80.00   
StockOptionLevel          1470.0      0.793878     0.852077     0.0     0.00   
TotalWorkingYears         1470.0     11.279592     7.780782     0.0     6.00   
TrainingTimesLastYear     1470.0      2.799320     1.289271     0.0     2.00   
WorkLifeBalance           1470.0      2.761224     0.706476     1.0     2.00   
YearsAtCompany            1470.0      7.008163     6.126525     0.0     3.00   
YearsInCurrentRole        1470.0      4.229252     3.623137     0.0     2.00   
YearsSinceLastPromotion   1470.0      2.187755     3.222430     0.0     0.00   
YearsWithCurrManager      1470.0      4.123129     3.568136     0.0     2.00   

                              50%       75%      max  
Age                          36.0     43.00     60.0  
DailyRate                   802.0   1157.00   1499.0  
DistanceFromHome              7.0     14.00     29.0  
Education                     3.0      4.00      5.0  
EmployeeCount                 1.0      1.00      1.0  
EmployeeNumber             1020.5   1555.75   2068.0  
EnvironmentSatisfaction       3.0      4.00      4.0  
HourlyRate                   66.0     83.75    100.0  
JobInvolvement                3.0      3.00      4.0  
JobLevel                      2.0      3.00      5.0  
JobSatisfaction               3.0      4.00      4.0  
MonthlyIncome              4919.0   8379.00  19999.0  
MonthlyRate               14235.5  20461.50  26999.0  
NumCompaniesWorked            2.0      4.00      9.0  
PercentSalaryHike            14.0     18.00     25.0  
PerformanceRating             3.0      3.00      4.0  
RelationshipSatisfaction      3.0      4.00      4.0  
StandardHours                80.0     80.00     80.0  
StockOptionLevel              1.0      1.00      3.0  
TotalWorkingYears            10.0     15.00     40.0  
TrainingTimesLastYear         3.0      3.00      6.0  
WorkLifeBalance               3.0      3.00      4.0  
YearsAtCompany                5.0      9.00     40.0  
YearsInCurrentRole            3.0      7.00     18.0  
YearsSinceLastPromotion       1.0      3.00     15.0  
YearsWithCurrManager          3.0      7.00     17.0

## insights:


1. **Age Distribution**: The average age of employees in the dataset is approximately 36.9 years, with a standard deviation of 9.14. The ages range from 18 to 60 years, with the majority falling between 30 and 43 years old.

2. **Daily Rate**: The average daily rate is around $802.49, with a standard deviation of $403.51. Daily rates vary widely, ranging from $102 to $1499.

3. **Distance From Home**: On average, employees live approximately 9.19 units of distance from their workplace, with a standard deviation of 8.11. The distance ranges from 1 to 29 units.

4. **Education Level**: The average education level is 2.91, indicating that most employees have completed some college or technical training. Education levels range from 1 (high school) to 5 (doctorate).

5. **Monthly Income**: The average monthly income is $6502.93, with a significant standard deviation of $4707.96. Monthly incomes vary widely, ranging from $1009 to $19999.

6. **Job Satisfaction and Engagement**: Employees generally report moderate levels of job satisfaction and engagement. The average scores for job satisfaction, environment satisfaction, job involvement, and relationship satisfaction are around 2.72 to 2.73, on a scale of 1 to 4.

7. **Job Level and Experience**: The average job level is 2.06, indicating that most employees hold mid-level positions. Total working years average around 11.28, with a wide standard deviation of 7.78, suggesting a diverse range of experience levels within the dataset.

8. **Promotions and Managerial Relationships**: On average, employees have been with their current manager for approximately 4.12 years. However, years since the last promotion average only 2.19, indicating potential opportunities for career advancement.



In [59]:
df.describe(include='O')

Attrition BusinessTravel              Department EducationField Gender  \
count       1470           1470                    1470           1470   1470   
unique         2              3                       3              6      2   
top           No  Travel_Rarely  Research & Development  Life Sciences   Male   
freq        1233           1043                     961            606    882   

                JobRole MaritalStatus Over18 OverTime  
count              1470          1470   1470     1470  
unique                9             3      1        2  
top     Sales Executive       Married      Y       No  
freq                326           673   1470     1054

In [60]:
# is there any null values present in the dataset
df.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

In [61]:
# is there any duplicated values present in the dataset
df.duplicated().sum()

0

# **5. Data Preprocessing**

# 5.1 Data Cleaning

In [62]:
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   
1                        3    Male          61               2         2   
2                        4    Male          92               2         1   
3                        4  Female          56               3         1   
4                        1    Male          40               3         1   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  \
0        Sales Executive                4        Single           5993   
1     Research Scientist                2       Married           5130   
2  Laboratory Technician                3        Single           2090   
3     Research Scientist                3       Married           2909   
4  Laboratory Technician                2       Married           3468   

   MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
0        19479                   8      Y      Yes                 11   
1        24907                   1      Y       No                 23   
2         2396                   6      Y      Yes                 15   
3        23159                   1      Y      Yes                 11   
4        16632                   9      Y       No                 12   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
0                  3                         1             80   
1                  4                         4             80   
2                  3                         2             80   
3                  3                         3             80   
4                  3                         4             80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [63]:
df.drop(columns=['Over18'], inplace=True)

In [64]:
df['Attrition'].value_counts()

No     1233
Yes     237
Name: Attrition, dtype: int64

In [65]:
# checking for null values
df.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSinceLastPromotion     0
YearsWithC

In [66]:
# Duplicated values
df.duplicated().sum()

0

# 5.2 Split Data Into X and y:

In [67]:
# split data into X and y

X = df.drop('Attrition',axis=1)
y = df.Attrition

In [68]:
from sklearn.model_selection import train_test_split

from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OrdinalEncoder,OneHotEncoder
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import MinMaxScaler

from sklearn.feature_selection import SelectKBest,chi2

from sklearn.pipeline import Pipeline

from sklearn.ensemble import RandomForestRegressor

from imblearn.over_sampling import SMOTE

In [69]:
# training data and testing data

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [70]:
pd.set_option('display.max_columns',None)

In [71]:
X_train

Age     BusinessTravel  DailyRate              Department  \
1097   24      Travel_Rarely        350  Research & Development   
727    18         Non-Travel        287  Research & Development   
254    29      Travel_Rarely       1247                   Sales   
1175   39      Travel_Rarely        492  Research & Development   
1341   31      Travel_Rarely        311  Research & Development   
...   ...                ...        ...                     ...   
1130   35      Travel_Rarely        750  Research & Development   
1294   41      Travel_Rarely        447  Research & Development   
860    22  Travel_Frequently       1256  Research & Development   
1459   29      Travel_Rarely       1378  Research & Development   
1126   50      Travel_Rarely        264                   Sales   

      DistanceFromHome  Education    EducationField  EmployeeCount  \
1097                21          2  Technical Degree              1   
727                  5          2     Life Sciences              1   
254                 20          2         Marketing              1   
1175                12          3           Medical              1   
1341                20          3     Life Sciences              1   
...                ...        ...               ...            ...   
1130                28          3     Life Sciences              1   
1294                 5          3     Life Sciences              1   
860                  3          4     Life Sciences              1   
1459                13          2             Other              1   
1126                 9          3         Marketing              1   

      EmployeeNumber  EnvironmentSatisfaction Gender  HourlyRate  \
1097            1551                        3   Male          57   
727             1012                        2   Male          73   
254              349                        4   Male          45   
1175            1654                        4   Male          66   
1341            1881                        2   Male          89   
...              ...                      ...    ...         ...   
1130            1596                        2   Male          46   
1294            1814                        2   Male          85   
860             1203                        3   Male          48   
1459            2053                        4   Male          46   
1126            1591                        3   Male          59   

      JobInvolvement  JobLevel                    JobRole  JobSatisfaction  \
1097               2         1      Laboratory Technician                1   
727                3         1         Research Scientist                4   
254                3         2            Sales Executive                4   
1175               3         2     Manufacturing Director                2   
1341               3         2      Laboratory Technician                3   
...              ...       ...                        ...              ...   
1130               4         2      Laboratory Technician                3   
1294               4         2  Healthcare Representative                2   
860                2         1         Research Scientist                4   
1459               2         2      Laboratory Technician                2   
1126               3         5                    Manager                3   

     MaritalStatus  MonthlyIncome  MonthlyRate  NumCompaniesWorked OverTime  \
1097      Divorced           2296        10036                   0       No   
727         Single           1051        13493                   1       No   
254       Divorced           6931        10732                   2       No   
1175       Married           5295         7693                   4       No   
1341      Divorced           4197        18624                   1       No   
...            ...            ...          ...                 ...      ...   
1130       Married           3407        25348       

In [72]:
# Encoding the target column "Attrition"
y_train = y_train.map({'Yes': 1, 'No': 0})
y_test = y_test.map({'Yes': 1, 'No': 0})

In [73]:
y_train

1097    0
727     0
254     0
1175    0
1341    0
       ..
1130    0
1294    0
860     1
1459    0
1126    0
Name: Attrition, Length: 1176, dtype: int64

In [74]:
X_train['OverTime'].value_counts()

No     837
Yes    339
Name: OverTime, dtype: int64

In [75]:
# column transformer for encoding categorical columns

encoder = ColumnTransformer(
    transformers=[
        ('ordinal', OrdinalEncoder(), ['BusinessTravel', 'Department', 'EducationField', 'JobRole', 'OverTime']),  # Ordinal encoding
        ('ohe', OneHotEncoder(drop='first', sparse_output=False), ['Gender', 'MaritalStatus']),  # One-hot encoding with drop_first=True for 'Gender'
    ],
    remainder='passthrough'  # Keep the other columns unchanged
)

# setting to get a pandas df
encoder.set_output(transform='pandas')

ColumnTransformer(remainder='passthrough',
                  transformers=[('ordinal', OrdinalEncoder(),
                                 ['BusinessTravel', 'Department',
                                  'EducationField', 'JobRole', 'OverTime']),
                                ('ohe',
                                 OneHotEncoder(drop='first',
                                               sparse_output=False),
                                 ['Gender', 'MaritalStatus'])])

In [76]:
# Define the pipeline
pipe = Pipeline([
    ('preprocessing', encoder),  # Assuming `encoder` is your previously defined encoder
    ('scaling', MinMaxScaler()),  # Scaling step
    ('feature_selection', SelectKBest(score_func=chi2, k=15)),  # Feature selection step
])

# Fit the pipeline on the training data
pipe.fit(X_train, y_train)

Pipeline(steps=[('preprocessing',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('ordinal', OrdinalEncoder(),
                                                  ['BusinessTravel',
                                                   'Department',
                                                   'EducationField', 'JobRole',
                                                   'OverTime']),
                                                 ('ohe',
                                                  OneHotEncoder(drop='first',
                                                                sparse_output=False),
                                                  ['Gender',
                                                   'MaritalStatus'])])),
                ('scaling', MinMaxScaler()),
                ('feature_selection',
                 SelectKBest(k=15,
                             score_func=<function chi2 at 0x0000017E2F6DD040>))])

In [77]:
import pickle

# Assuming `pipe` is your fitted pipeline
# Save the pipeline to a file
with open('preprocessing_pipeline.pkl', 'wb') as f:
    pickle.dump(pipe, f)


In [78]:
# Transform both the training and testing data
X_train_transformed = pd.DataFrame(pipe.transform(X_train))
X_test_transformed = pd.DataFrame(pipe.transform(X_test))

In [79]:
X_train_transformed

0    1    2         3         4         5         6     7         8   \
0     0.0  0.0  0.0  0.142857  0.714286  0.666667  0.333333  0.00  0.000000   
1     0.0  0.0  1.0  0.000000  0.142857  0.333333  0.666667  0.00  1.000000   
2     0.0  0.0  0.0  0.261905  0.678571  1.000000  0.666667  0.25  1.000000   
3     0.0  1.0  0.0  0.500000  0.392857  1.000000  0.666667  0.25  0.333333   
4     0.0  0.0  0.0  0.309524  0.678571  0.333333  0.666667  0.25  0.666667   
...   ...  ...  ...       ...       ...       ...       ...   ...       ...   
1171  0.0  1.0  0.0  0.404762  0.964286  0.333333  1.000000  0.25  0.666667   
1172  0.0  0.0  1.0  0.547619  0.142857  0.333333  1.000000  0.25  0.333333   
1173  1.0  1.0  0.0  0.095238  0.071429  0.666667  0.333333  0.00  1.000000   
1174  1.0  1.0  0.0  0.261905  0.428571  1.000000  0.333333  0.25  0.333333   
1175  1.0  1.0  0.0  0.761905  0.285714  0.666667  0.666667  1.00  0.666667   

            9         10     11     12        13        14  
0     0.067773  1.000000  0.050  0.025  0.055556  0.000000  
1     0.002212  0.000000  0.000  0.000  0.000000  0.000000  
2     0.311848  0.333333  0.250  0.075  0.111111  0.117647  
3     0.225698  0.000000  0.175  0.125  0.222222  0.000000  
4     0.167878  0.333333  0.250  0.250  0.444444  0.117647  
...        ...       ...    ...    ...       ...       ...  
1171  0.126277  0.666667  0.250  0.250  0.500000  0.470588  
1172  0.308636  0.000000  0.275  0.075  0.111111  0.117647  
1173  0.097104  0.333333  0.025  0.000  0.000000  0.000000  
1174  0.158820  0.333333  0.250  0.100  0.166667  0.176471  
1175  0.964824  0.333333  0.675  0.025  0.000000  0.000000  

[1176 rows x 15 columns]

In [80]:
sm = SMOTE()
x_smote, y_smote = sm.fit_resample(X_train_transformed, y_train)

In [81]:
y_smote.value_counts()

0    978
1    978
Name: Attrition, dtype: int64

# **6. Model Building**

# 6.1 Decision Tree

In [82]:
from sklearn.tree import DecisionTreeClassifier

from sklearn.metrics import accuracy_score,classification_report,f1_score

In [83]:
model = DecisionTreeClassifier()
model.fit(x_smote, y_smote)

# prediction with x_test
y_pred=model.predict(X_test_transformed)
#prediction with x_train
y_train_predict=model.predict(X_train_transformed)

In [84]:
# training accuracy
accuracy_score(y_train,y_train_predict)

1.0

In [85]:
# testing accuracy
accuracy_score(y_test,y_pred)

0.7687074829931972

* we see that Model is overfitting beacause training score is high and testing score is less.
* To overcome overfitting we do hyperparmeter tunning

# Hyper-parameter tuning

In [86]:
from sklearn.model_selection import GridSearchCV
params = {
    "criterion":("gini", "entropy"),
    "splitter":("best", "random"),
    "max_depth":(list(range(1, 20))),
    "min_samples_split":[2, 3, 4],
    "min_samples_leaf":list(range(1, 20)),
}

In [87]:
tree_clf = DecisionTreeClassifier(random_state=3)
tree_cv = GridSearchCV(tree_clf, params, scoring="accuracy", n_jobs=-1, verbose=1, cv=3)
tree_cv.fit(x_smote,y_smote)
best_params = tree_cv.best_params_
print(f"Best paramters: {best_params})")

Fitting 3 folds for each of 4332 candidates, totalling 12996 fits


Best paramters: {'criterion': 'gini', 'max_depth': 15, 'min_samples_leaf': 1, 'min_samples_split': 2, 'splitter': 'random'})


In [88]:
tree_cv.best_score_

0.8640081799591002

In [89]:
model=DecisionTreeClassifier(criterion='entropy',max_depth=14,min_samples_leaf= 1,min_samples_split=2,splitter='random')
model.fit(x_smote,y_smote)#training the model
y_pred=model.predict(X_test_transformed)#prediction
y_pred#predicted values

array([0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0], dtype=int64)

In [90]:
y_train_predict=model.predict(x_smote)#predicting training data to check training performance
y_train_predict

array([0, 0, 0, ..., 1, 1, 1], dtype=int64)

In [91]:
# Training score
accuracy_score(y_smote,y_train_predict)

0.9749488752556237

In [92]:
import pickle

In [93]:
Pkl_Filename = "attrition_prediction_model.pkl"  

with open(Pkl_Filename, 'wb') as file:  
    pickle.dump(model, file)

In [94]:
import pickle

# Extract encoding mappings from the OneHotEncoder transformer
ohe_transformer = encoder.named_transformers_['ohe']
categories = ohe_transformer.categories_

# Create a dictionary to store the encoding mappings
encoding_mappings = {
    'categories': categories,
}

for i in categories:
    print(i)


['Female' 'Male']
['Divorced' 'Married' 'Single']
